splinepy/helpme/fit.py

```diff
@@ -0,0 +1,738 @@
+        # interpolation
+        # NURBS Book eq. 9.8 (n_points = m + 1, n_control_points = n + 1)
+        knot_vector[(degree + 1) : -(degree + 1)] = (
+            _np.convolve(u_k.ravel(), _np.ones(degree), "valid")[1:-1] / degree
+        )
+
+    else:
+        # approximation
+        # NURBS Book eq. (9.68) (n_points = m + 1, n_control_points = n + 1)
+        d = (n_points) / (n_control_points - degree)
+        # NURBS Book eq. (9.69)
+        for j in range(1, n_control_points - degree):
+            i = int(j * d)
+            alpha = (j * d) - i
+            knot_vector[j + degree] = (1 - alpha) * u_k[i - 1] + alpha * u_k[i]
+    return knot_vector
+
+
+def solve_for_control_points(
+    target_points, target_spline, queries, interpolate_endpoints
+):
+    """
+    Builds and solves the system to calculate control points.
+
+    Parameters
+    ----------
+    target_points: (m + 1 x dim) array
+        points to be interpolated/approximated
+
+    target_spline: spline
+        spline used for interpolation/approximation
+
+    queries: (n x 1) array
+        values where the spline will be evaluated
+
+    interpolate_endpoints: bool
+        if True -> endpoints are interpolated
+
+    Returns
+    -------
+    residual: float
+        residual (coefficient_matrix @ control_points - points)
+    """
+    # build matrix
+    coefficient_matrix = _make_matrix(
+        *target_spline.basis_and_support(queries),
+        target_spline.control_points.shape[0],
+    )
+    points = target_points.copy()
+
+    if (
+        interpolate_endpoints
+        and target_spline.control_points.shape != points.shape
+    ):
+        # reduction of matrix necessary for interpolation of endpoints
+        target_spline.control_points[0] = points[0]
+        # [-1] index doesn't work -> therefore actual index is used
+        target_spline.control_points[
+            target_spline.control_points.shape[0] - 1
+        ] = points[-1]
+
+        if target_spline.control_points.shape[0] == 2:
+            # control points = endpoints
+            residual = _np.linalg.norm(
+                coefficient_matrix @ target_spline.control_points - points
+            )
+        elif _has_scipy:
+            # move known values to RHS
+            points[1:-1, :] += (
+                -coefficient_matrix[1:-1, [0]] * points[0, :]
+                - coefficient_matrix[1:-1, [-1]] * points[-1, :]
+            ).todense()
+
+            # solve system A^TAx=A^Tb
+            at = coefficient_matrix[1:-1, 1:-1].transpose(copy=True)
+            target_spline.control_points[1:-1] = _spsolve(
+                at @ coefficient_matrix[1:-1, 1:-1], at @ points[1:-1]
+            )
+
+            # alternative using lsqr
+
+            # for i in range(0, points.shape[1]):
+            #     (
+            #         target_spline.control_points[1:-1, i]
+            #     ) = _lsqr(coefficient_matrix[1:-1, 1:-1], points[1:-1, i])[0]
```

Comment:
Please delete all unused code If at all, leave a comment saying that you have tried using lsqr but it showed worse performance for future developers